src/agent/thread_helper.cpp

```diff
@@ -0,0 +1,323 @@
+                          const std::vector<uint8_t> &aMasterKey,
+                          const std::vector<uint8_t> &aPSKc,
+                          uint32_t                    aChannelMask,
+                          const OtResultHandler &     aHandler)
+
+{
+    otError         err = OT_ERROR_NONE;
+    otExtendedPanId extPanId;
+    otMasterKey     masterKey;
+    otPskc          pskc;
+    otError         pskcRet;
+    uint32_t        channelMask;
+    uint8_t         channel;
+
+    VerifyOrExit(mAttachHandler == nullptr && mJoinerHandler == nullptr, err = OT_ERROR_INVALID_STATE);
+    mAttachHandler = std::move(aHandler);
+    if (aPanId == UINT16_MAX)
+    {
+        RandomFill(&aPanId, sizeof(aPanId));
+    }
+
+    if (aExtPanId != UINT64_MAX)
+    {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+        uint8_t *srcBuffer = reinterpret_cast<uint8_t *>(&aExtPanId);
+
+        for (size_t i = 0; i < sizeof(extPanId.m8); i++)
+        {
+            extPanId.m8[i] = srcBuffer[sizeof(extPanId.m8) - 1 - i];
+        }
+#else
+        memcpy(extPanId.m8, &aExtPanId, sizeof(extPanId.m8));
+#endif // __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    }
+    else
+    {
+        RandomFill(extPanId.m8, sizeof(extPanId.m8));
+    }
+
+    VerifyOrExit(aMasterKey.empty() || aMasterKey.size() == sizeof(masterKey.m8), err = OT_ERROR_INVALID_ARGS);
+    if (!aMasterKey.empty())
+    {
+        memcpy(masterKey.m8, &aMasterKey[0], sizeof(masterKey.m8));
+    }
+    else
+    {
+        RandomFill(masterKey.m8, sizeof(masterKey.m8));
+    }
+
+    VerifyOrExit(aPSKc.empty() || aPSKc.size() == sizeof(pskc.m8), err = OT_ERROR_INVALID_ARGS);
+    if (!aPSKc.empty())
+    {
+        memcpy(pskc.m8, &aPSKc[0], sizeof(pskc.m8));
+    }
+    else
+    {
+        RandomFill(pskc.m8, sizeof(pskc.m8));
+    }
+
+    SuccessOrExit(otThreadSetNetworkName(mInstance, aName.c_str()));
+    SuccessOrExit(otLinkSetPanId(mInstance, aPanId));
+    SuccessOrExit(otThreadSetExtendedPanId(mInstance, &extPanId));
+    SuccessOrExit(otThreadSetMasterKey(mInstance, &masterKey));
+
+    channelMask = otPlatRadioGetPreferredChannelMask(mInstance) & aChannelMask;
+
+    if (channelMask == 0)
+    {
+        channelMask = otLinkGetSupportedChannelMask(mInstance) & aChannelMask;
+    }
+
+    VerifyOrExit(channelMask != 0, err = OT_ERROR_INVALID_ARGS);
+    channel = RandomChannelFromChannelMask(channelMask);
+    SuccessOrExit(otLinkSetChannel(mInstance, channel));
+
+    pskcRet = otThreadSetPskc(mInstance, &pskc);
+    VerifyOrExit(pskcRet == OT_ERROR_NONE || pskcRet == OT_ERROR_NOT_IMPLEMENTED, err = OT_ERROR_FAILED);
+
+    if (!otIp6IsEnabled(mInstance))
+    {
+        SuccessOrExit(err = otIp6SetEnabled(mInstance, true));
+    }
+    SuccessOrExit(err = otThreadSetEnabled(mInstance, true));
+exit:
+    if (err != OT_ERROR_NONE)
+    {
+        mAttachHandler(err);
+        mAttachHandler = nullptr;
+    }
+}
+
+void ThreadHelper::Leave(const OtResultHandler &aHandler)
```

Comment:
```suggestion
void ThreadHelper::Leave(OtResultHandler aHandler)
```
Moving a const reference doesn't make sense.

Reply:
+1